### PR TITLE
Support TypeScript 4.7

### PR DESCRIPTION
**Pull request by Dionysusnu**

### CODE REVIEW

**Comment by Dionysusnu** [comment]
Ah, GitHub does not like me renaming the branch to `support-typescript-4.7`..

**Comment by Dionysusnu** [comment]
CI will fail for a bit because ts-expose-internals trails behind TS releases by about a day. Keeping PR as draft for that reason as well.

File: package.json [hunk 1]
```diff
@@ -58,7 +58,7 @@
 		"lookpath": "^1.2.2",
 		"prompts": "^2.4.1",
 		"resolve": "^1.20.0",
-		"typescript": "=4.6.2",
+		"typescript": "^4.7.0",
```
Comment (Dionysusnu) [inline package.json]: Updating the TS version

File: package.json [hunk 2]
```diff
@@ -78,8 +79,7 @@
 		"mocha": "^9.1.1",
 		"nyc": "^15.1.0",
 		"prettier": "^2.4.1",
-		"ts-expose-internals": "=4.5.5",
```
Comment (Dionysusnu) [inline package.json:81]: Moving ts-expose-internals to its new recommended install procedure. Doing this means that the line removed from `tsconfig-base.json` is redundant.

File: package.json [hunk 3]
```diff
@@ -78,8 +79,7 @@
...
-		"ttypescript": "^1.5.12",
+		"ttypescript": "^1.5.13",
```
Comment (Dionysusnu) [inline package.json:82]: Didn't mean to do this, probably got done by npm somewhere in the conflict resolution with the beta versions. I guess it's fine to stay?

File: tests/src/tests/map.spec.ts [hunk 4]
```diff
@@ -120,7 +120,7 @@ export = () => {
 		]);
 		const map2 = new Map<string, number>(arr);
 
-		function values<V>(map: Map<unknown, V>): Array<V> {
+		function values<V extends defined>(map: Map<unknown, V>): Array<V> {
```
Comment (Dionysusnu) [inline tests/src/tests/map.spec.ts:123]: Breaking typechecker change in TS 4.7, generic types must now extend the class generics too. For Map here, that means it needs to extend defined.

File: tests/package.json [hunk 5]
```diff
@@ -10,10 +10,10 @@
 	"author": "",
 	"license": "ISC",
 	"dependencies": {
-		"@rbxts/compiler-types": "^1.3.2-types.0",
+		"@rbxts/compiler-types": "^1.3.3-types.1",
 		"@rbxts/roact": "^1.3.0-ts.7",
 		"@rbxts/services": "^1.1.2",
 		"@rbxts/testez": "^0.3.1-ts.6",
-		"@rbxts/types": "^1.0.571"
+		"@rbxts/types": "^1.0.589"
```
Comment (Dionysusnu) [inline tests/package.json:17]: Just updating the test types (done automatically on running `npm test`)

**Review by Dionysusnu — COMMENTED** [review]
Summary of the changes included in code markings, or in the following list for the changes that span multiple files:
- With the new TS compilerOption `moduleDetection` set to `"force"`, files are always modules without the need for an import/export statement in them. This manually-enforced requirement has been removed in favour of enforcing the compilerOption.
- DiagnosticsProducingTypeChecker and TypeChecker classes were merged. They now act the same by just calling `program.getTypeChecker()`.

Still TODO: Remove the export statements from test files.

**Review by osyrisrblx — APPROVED** [review]

**Pull request merged**